base/set.jl

```diff
@@ -201,6 +201,76 @@ function unique(f::Callable, C)
     out
 end
 
+# If A is not sorted, then we will need to keep track of all of the elements that
+# we have seen so far.
+function _unique!(A)
+    seen = Set{eltype(A)}()
+    idxs = eachindex(A)
+    i = state = start(idxs)
+    for x in A
+        if x ∉ seen
+            push!(seen, x)
+            i, state = next(idxs, state)
+            A[i] = x
+        end
+    end
+    resize!(A, i - first(idxs) + 1)
+end
+
+# If A is sorted, then we only need to keep track of one element and add that to A
+# every time that we see a new element.
+function _sortedunique!(A)
+    idxs = eachindex(A)
+    y = first(A)
+    state = start(idxs)
+    j, state = next(idxs, state)
+    for i in idxs
+        x = A[i]
+        if !isequal(x, y)
+            j, state = next(idxs, state)
+            y = A[j] = x
+        end
+    end
+    resize!(A, j - first(idxs) + 1)
+end
+
+"""
+    unique!(A::AbstractVector)
+
+Remove duplicate items as determined by [`isequal`](@ref), then return the modified `A`.
+
+```jldoctest
+julia> unique!([1, 1, 1])
+1-element Array{Int64,1}:
+ 1
+
+julia> A = [7, 3, 2, 3, 7, 5];
+
+julia> unique!(A);
```

Comment:
Yes, `unique!` will return `A`, so for example
```julia
julia> A = [1,2,1,3,2];

julia> B = unique!(A)
3-element Array{Int64,1}:
 1
 2
 3

julia> B
3-element Array{Int64,1}:
 1
 2
 3
```
I have updated the docstring so that the semicolon is no longer there, so the user can see that `A` is being returned. I have also updated the tests so that they look like
```julia
u = [1,2,2,3,5,5]
@test unique!(u) === u
@test u == [1,2,3,5]
```
Therefore, the tests now check whether or not the returned value is correct.